segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -11,74 +11,75 @@
 import io.pravega.segmentstore.contracts.SegmentProperties;
 import io.pravega.segmentstore.contracts.StreamSegmentException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
 import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.StorageNotPrimaryException;
 import io.pravega.segmentstore.storage.SyncStorage;
+import java.io.EOFException;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
-import java.util.concurrent.Callable;
+import java.util.Arrays;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
 import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
+import lombok.val;
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
 import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsAction;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.io.IOUtils;
 
 /**
  * Storage adapter for a backing HDFS Store which implements fencing using file-chaining strategy.
  * <p>
- * Each segment is a collection of ordered files, adopting the following pattern: {segment-name}_{start-offset}_{epoch}.
+ * Each segment is represented by one file, adopting the following pattern: {segment-name}_{epoch}.
  * <ul>
  * <li> {segment-name} is the name of the segment as used in the SegmentStore
- * <li> {start-offset} is the offset within the Segment of the first byte in this file.
  * <li> {epoch} is the Container Epoch which had ownership of that segment when that file was created.
  * </ul>
  * <p>
  * Example: Segment "foo" can have these files
  * <ol>
- * <li> foo_0_1: first file, contains data at offsets [0..1234), written during Container Epoch 1.
- * <li> foo_1234_2: data at offsets [1234..987632), written during Container Epoch 2.
- * <li> foo_987632_3: data at offsets [987632..10568949), written during Container Epoch 3.
- * <li> foo_10568949_4: data at offsets [10568949..current-length-of-segment), written during Container Epoch 4.
- * </ol>
+ * <li> foo_<epoc>: Segement file, owned by a segmentstore running under epoch 1.
+ * <li> foo_<MAX_LONG>: A sealed segment.
  * <p>
- * All but the last file is marked as 'read-only' (except if the segment is Sealed, in which case the last file too is
- * marked as 'read-only').
+ * When a container fails over and needs to reacquire ownership of a segment, it renames the segment file as foo_<current_epoc>.
+ * After creation of the file, the filename is checked again. If there exists any file with higher epoc, the current file is deleted
```

Comment:
If the file is named foo_<number> and some other newer process has ranamed it to a higher number shouldn't there not be a file to delete? What is the file that is being deleted here?

Reply:
This is a scenario where two hosts are running the create code simultaneously. Both will see that the file does not exist and try to create it. The file is created with epoch so the hosts will end up creating two files `segment_1` and `segment_2`. If after creating the host with epoch `1` validates that there is no segment already exists with higher epoch, then we do not have a dangling file.
(We can also leave the file dangling too - it does not hurt as it is just an empty file).